ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/VolDB.hs

```diff
@@ -494,22 +514,36 @@ blockFileParser' :: forall m blk h. (IOLike m, HasHeader blk)
                  -> (blk -> IsEBB)
                  -> (blk -> BinaryInfo Encoding)
                  -> (forall s. Decoder s (Lazy.ByteString -> blk))
+                 -> (blk -> Bool)
                  -> VolDB.Parser
-                     Util.CBOR.ReadIncrementalErr
+                     (BlockFileParserError (HeaderHash blk))
                      m
                      (HeaderHash blk)
-blockFileParser' hasFS isEBB encodeBlock decodeBlock = VolDB.Parser $
-       fmap (fmap (fmap fst)) -- Drop the offset of the error
-     . Util.CBOR.readIncrementalOffsets hasFS decoder'
+blockFileParser' hasFS isEBB encodeBlock decodeBlock checkIntegrity =
+    VolDB.Parser $ \fsPath -> Util.CBOR.withStreamIncrementalOffsets
+      hasFS decodeBlock fsPath checkEntries
   where
     -- TODO: It looks weird that we use an encoding function 'encodeBlock'
     -- during parsing, but this is quite cheap, since the encoding is already
     -- cached. We should consider improving this, so that it does not create
     -- confusion.
-    decoder' :: forall s. Decoder s (Lazy.ByteString
-             -> VolDB.BlockInfo (HeaderHash blk))
-    decoder' = ((\blk -> extractInfo isEBB (encodeBlock blk) blk) .)
-      <$> decodeBlock
+    extractInfo' :: blk -> VolDB.BlockInfo (HeaderHash blk)
+    extractInfo' blk = extractInfo isEBB (encodeBlock blk) blk
+
+    checkEntries :: Stream (Of (Word64, (Word64, blk)))
+                    m
+                    (Maybe (Util.CBOR.ReadIncrementalErr, Word64))
+                 -> m (VolDB.ParsedInfo (HeaderHash blk),
+                    Maybe (BlockFileParserError (HeaderHash blk)))
+    checkEntries = go []
+      where
+        go parsed stream = S.next stream >>= \case
+          Left mbErr -> return (reverse parsed, BlockReadErr . fst <$> mbErr)
+          Right ((offset, (size, blk)), stream') | checkIntegrity blk ->
+            let newParsed = (offset, (VolDB.BlockSize size, extractInfo' blk))
```

Comment:
We have to be careful here to not accidentally hold on to the block. As we're returning a tuple, we can't rely on the data type to do the forcing for us. You can do the following:
```haskell
let !blockInfo = extractInfo' blk
    newParsed  = (offset, (VolDB.BlockSize size, blockInfo))
...
```